create-image

```diff
@@ -0,0 +1,422 @@
+		Linux)
+			mount "$file" "$mntpt" -o loop
+			;;
+		SunOS)
+			mount -F hsfs "$file" "$mntpt"
+			;;
+		*)
+			echo "$0: do not know how to mount an iso on" \
+			    "$(uname -s)". 1>&2
+			exit 2
+			;;
+	esac
+}
+
+create_ks_root() {
+	local src_mntpt=/tmp/mnt/$distrel
+
+	echo "Creating ISO Layout"
+
+	if [[ -d $ks_root ]]; then
+		echo "Layout $ks_root exists...nuking"
+		rm -rf "$ks_root"
+	fi
+	echo "Creating $ks_root"
+	mkdir -p "$ks_root"
+
+	if [[ -d $custom_rpms ]]; then
+		echo "Copying custom RPMS"
+		find "$custom_rpms" -type f -exec cp {} "$ks_root/Packages" \;
+	fi
+
+	if [[ -f $custom_rpms.remote ]]; then
+		echo "Downloading custom RPMS"
+		local rpm
+		mkdir -p "$ks_root/Packages"
+		pushd "$ks_root/Packages" >/dev/null
+		for rpm in $(sed 's/#.*//' $custom_rpms.remote); do
+			echo "  $(basename $rpm)"
+			curl -O "$rpm"
+		done
+		popd >/dev/null
+	fi
+
+	echo "Finished Populating Layout"
+}
+
+copy_ks_cfg() {
+	echo "Copying Kickstart file"
+	cp $ks_cfg $ks_root/ks.cfg
+}
+
+copy_guest_tools() {
+	echo "Copying $guesttools"
+	echo "Initiallizing and fetching submodule $guesttools"
+	git submodule update --init
+	cp -R ./$guesttools/ $ks_root/
+}
+
+create_ks_iso() {
+	echo "Preparing kickstart ISO"
+	copy_guest_tools
+	copy_ks_cfg
+	echo "Creating kickstart ISO"
+	mkisofs -r -R -J -T -v -V "kickstart" -A "kickstart" -o $ks_iso $ks_root
+	echo "Kickstart $ks_iso is ready"
+}
+
+create_image() {
+	local desc
+	local media_mntpt=/tmp/mnt/$distrel-media
+	local qemu_args=()
+	local append=
+
+	get_iso_file_sha
+
+	# create-hybrid-image puts the boot iso at index=0
+	qemu_args+=( "-drive" "file=$ks_iso,if=ide,index=1,media=cdrom" )
+
+	#
+	# We must specify boot args to initiate an unattended installation.
+	#
+	# Boot messages to serial console (which is on stdout), debugging and
+	# interaction with the installer possible on the VNC console.  A %pre
+	# command in the kickstart file will tail install logs to /dev/ttyS0
+	append="console=ttyS0 console=tty0"
+	# Ensure that network device names are predictable
+	append+=" net.ifnames=0 biosdevname=0"
+	# The install media is on the first dvd
+	append+=" ${BOOT_INST_PREFIX}stage2=hd:/dev/sr0"
+	# The kickstart config is on the second dvd
+	append+=" ${BOOT_INST_PREFIX}ks=hd:/dev/sr1:/ks.cfg"
+	# Guest boot scripts run 'set -x' when tracing. Ignored before CentOS 7.
+	[[ -n "$TRACE" ]] && append+=" rd.debug"
+
+	qemu_args+=( "-append" "$append" )
+
+	#
+	# In order to specify boot args, we have to specify the kernel, initrd,
+	# and root file system
+	#
+	mount_iso "$iso_dir/$iso_file" "$media_mntpt"
+	qemu_args+=( "-kernel" "$media_mntpt/images/pxeboot/vmlinuz" )
+	qemu_args+=( "-initrd" "$media_mntpt/images/pxeboot/initrd.img" )
+
+	desc="$distro $release 64-bit image with just essential packages "
+	desc+="installed. Built to run on bhyve or KVM virtual machines."
+
+	local stamp=$branch-${cur_time//[:-]}-$githash
+	$top/eng/tools/create-hybrid-image -i "$iso_dir/$iso_file" \
```

Comment:
I was wondering what the centos equivalent of partman pre-seed magic was.  Looks like it's here?

Reply:
Mainly, create-hybrid-image does the layout for this case, right?

Reply:
CentOS has sane defaults and didn't require the manual partition layout voodoo that is required with Debian.  On CentOS, cloud-init requires cloud-utils-growpart (see ks.cfg) to do the resize.  This package doesn't exist on CentOS 6 and I'm not going to worry about that.  People really should not be deploying new CentOS 6 instances, as it goes from LTS to EOL in November of this year.